File: eduflow/backend/routers/comentarios.py
from fastapi import APIRouter, HTTPException, Query
from bson import ObjectId
from datetime import datetime
from pymongo import ReturnDocument

from database import get_db
from models.comentario import ComentarioCreate, RespuestaCreate
from utils import serialize_doc

router = APIRouter()


def _oid(value: str) -> ObjectId:
    try:
        return ObjectId(value)
    except Exception:
        raise HTTPException(status_code=422, detail=f"ID inválido: {value}")


@router.get("")
async def list_comentarios(
    cursoId: str = Query(default=None),
    leccionId: str = Query(default=None),
):
    db = get_db()
    filt: dict = {}
    if cursoId:
        filt["cursoId"] = _oid(cursoId)
    if leccionId:
        filt["leccionId"] = _oid(leccionId)

    docs = await db["comentarios"].find(filt).sort("fecha", -1).to_list(length=500)
    return serialize_doc(docs)


@router.post("", status_code=201)
async def create_comentario(body: ComentarioCreate):
    db = get_db()
    doc = {
        "cursoId": _oid(body.cursoId),
        "usuarioId": _oid(body.usuarioId),
        "leccionId": _oid(body.leccionId),
        "contenido": body.contenido,
        "calificacion": body.calificacion,
        "fecha": datetime.utcnow(),
        "respuestas": [],
    }
    result = await db["comentarios"].insert_one(doc)
    return serialize_doc({**doc, "_id": result.inserted_id})


@router.post("/{comentario_id}/respuestas", status_code=201)
async def add_respuesta(comentario_id: str, body: RespuestaCreate):
    db = get_db()
    respuesta = {
        "usuarioId": _oid(body.usuarioId),
        "contenido": body.contenido,
        "fecha": datetime.utcnow(),
    }
    result = await db["comentarios"].find_one_and_update(
        {"_id": _oid(comentario_id)},
        {"$push": {"respuestas": respuesta}},
        return_document=ReturnDocument.AFTER,
    )
    if not result:
        raise HTTPException(status_code=404, detail="Comentario no encontrado")
    return serialize_doc(result)

File: eduflow/server/routes/comentarios.js
import { Router } from "express";
import { ObjectId } from "mongodb";
import { getDb } from "../db.js";

const router = Router();

// GET /api/comentarios?cursoId=&leccionId=
router.get("/", async (req, res) => {
  const { cursoId, leccionId } = req.query;
  const filter = {};
  if (cursoId) filter.cursoId = new ObjectId(cursoId);
  if (leccionId) filter.leccionId = new ObjectId(leccionId);

  try {
    const db = getDb();
    const comentarios = await db
      .collection("comentarios")
      .find(filter)
      .sort({ fecha: -1 })
      .toArray();
    res.json(comentarios);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// POST /api/comentarios
router.post("/", async (req, res) => {
  const { cursoId, usuarioId, leccionId, contenido, calificacion } = req.body;
  if (!cursoId || !usuarioId || !leccionId || !contenido)
    return res.status(400).json({ error: "cursoId, usuarioId, leccionId y contenido son requeridos" });

  try {
    const db = getDb();
    const comentario = {
      cursoId: new ObjectId(cursoId),
      usuarioId: new ObjectId(usuarioId),
      leccionId: new ObjectId(leccionId),
      contenido,
      calificacion: calificacion ?? null,
      fecha: new Date(),
      respuestas: [],
    };

    const result = await db.collection("comentarios").insertOne(comentario);
    res.status(201).json({ ...comentario, _id: result.insertedId });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// POST /api/comentarios/:id/respuestas
router.post("/:id/respuestas", async (req, res) => {
  const { usuarioId, contenido } = req.body;
  if (!usuarioId || !contenido)
    return res.status(400).json({ error: "usuarioId y contenido son requeridos" });

  try {
    const db = getDb();
    const respuesta = {
      usuarioId: new ObjectId(usuarioId),
      contenido,
      fecha: new Date(),
    };

    const result = await db.collection("comentarios").findOneAndUpdate(
      { _id: new ObjectId(req.params.id) },
      { $push: { respuestas: respuesta } },
      { returnDocument: "after" }
    );

    if (!result) return res.status(404).json({ error: "Comentario no encontrado" });
    res.status(201).json(result);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

export default router;
